### PR TITLE
Simplified Onboarding Flow

### PR DESCRIPTION
**Pull request by philippzagar**
# Simplified Onboarding Flow

## :recycle: Current situation & Problem
The current implementation of the onboarding flow in our Template Application encompasses multiple stages which involve informing users, obtaining their consent, and requesting permissions for various iOS features, such as notifications and HealthKit. However, these stages do not consistently appear for every user. For example, if permissions for notifications and HealthKit are pre-existing, those specific onboarding steps are omitted.

The issue lies in the fact that the Spezi framework currently offers minimal infrastructure to streamline the construction of such a multi-faceted onboarding process. While it does provide a wide array of reusable views, it does not accommodate the complexity of an elaborate onboarding flow that often exists in production apps. This complexity necessitates the inclusion of custom logic from the application side to manage the onboarding flow, such as condition-based skipping of steps.

## :bulb: Proposed solution
Utilizing the new onboarding flow APIs introduced in https://github.com/StanfordSpezi/SpeziOnboarding/pull/10, this PR radically simplifies the template application's onboarding process by deleting repeated navigation-concerned conditionals as well as associated properties. This is achieved by using the `OnboardingStack` and the `OnboardingNavigationPath` from the `SpeziOnboarding` module.

## :gear: Release Notes 
- Template application utilizes simplified onboarding flow API of `SpeziOnboarding` 

## :heavy_plus_sign: Additional Information
The integration of the `SpeziOnboarding` flow API has significantly streamlined and improved the clarity of the onboarding code in the template application. However, one notable drawback is the increased overhead for the Preview Simulator views in Xcode due to the inability to share state across multiple previews, coupled with additional constraints related to `HealthKit` and other factors.
To address this challenge, a parameter named `stepToStart` has been introduced to the `OnboardingStack` API within `SpeziOnboarding`. This parameter designates the initial onboarding step, thereby enhancing the functionality for previews.
However, a limitation that remains is that previews necessitate a static list of all onboarding views. While this approach is not ideal, I couldn't think of a better and still "simple" one. Feel free to give me some suggestions here.

### Related PRs
`SpeziOnboarding` module PR: https://github.com/StanfordSpezi/SpeziOnboarding/pull/10

### Testing
UI tests are properly adjusted.

### Reviewer Nudging
Not too much new code, mostly the application of APIs from `SpeziOnboarding` to simplify the onboarding flow as well as some not so pretty stuff for the previews.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by PSchmiedmayer — CHANGES_REQUESTED** [review]
Thank you for showcasing how the changes here would affect the template app, I like the reduction in complexity and lines of code. I added some initial thoughts to the PR in the Onboarding Module that will translate to some changes here: https://github.com/StanfordSpezi/SpeziOnboarding/pull/10#pullrequestreview-1520658327

**Comment by philippzagar** [comment]
Of course, I will change the used `SpeziOnboarding` dependency back to a versioned release as soon as https://github.com/StanfordSpezi/SpeziOnboarding/pull/10 is merged :)

File: TemplateApplication/Helper/ProcessInfo+PreviewSimulator.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Helper/ProcessInfo+PreviewSimulator.swift:10]: Slightly nitpicky, but would be great to keep that convention around for all projects as noted in the Onboarding PR: https://github.com/StanfordSpezi/SpeziOnboarding/pull/10#discussion_r1285266106. 
```suggestion


```

Applies to some other Swift files as well 👍

File: TemplateApplication/Onboarding/AccountSetup/IconView.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/AccountSetup/IconView.swift:13]: 👍

File: TemplateApplication/Onboarding/Consent.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/Consent.swift:39]: That is just nice to see as a diff 🚀

File: TemplateApplication/Onboarding/HealthKitPermissions.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/HealthKitPermissions.swift:62]: Thank you!

File: TemplateApplication/Onboarding/InterestingModules.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/InterestingModules.swift:41]: As we moved this to the main OnboardingStack View Builder, we could remove this check here, couldn't we?

File: TemplateApplication/Onboarding/NotificationPermissions.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/NotificationPermissions.swift:53]: I like that this is now done automatically, this will create nice diffs if we append anything to the onboarding flow. I remember that this was a pain point in the @CS342 class last year.

File: TemplateApplication/Onboarding/OnboardingFlow.swift
Comment (PSchmiedmayer) [inline TemplateApplication/Onboarding/OnboardingFlow.swift:55]: This looks really nice, great job @philippzagar 🚀

**Review by PSchmiedmayer — APPROVED** [review]
Very nice job with the PR @philippzagar! I only had a few smaller comments that should be easy to address. Once these are addressed and the main onboarding PR is merged, you can merge this one without the need for a second review!

Thank you for the time and effort that went into all these improvements, this turned out really nice! 🚀

File: TemplateApplication/Onboarding/Consent.swift
Comment (philippzagar) [reply TemplateApplication/Onboarding/Consent.swift:39]: 🚀

File: TemplateApplication/Onboarding/InterestingModules.swift
Comment (philippzagar) [reply TemplateApplication/Onboarding/InterestingModules.swift:41]: You're completely right, missed that here.

File: TemplateApplication/Onboarding/OnboardingFlow.swift
Comment (philippzagar) [reply TemplateApplication/Onboarding/OnboardingFlow.swift:55]: 🚀

**Comment by philippzagar** [comment]
Thanks for the nice review @PSchmiedmayer! 🚀
One last thing: It seems like the UI tests are having issues on the runners right now, as can be seen here: https://github.com/StanfordSpezi/SpeziTemplateApplication/actions/runs/5779409484/job/15674435096
I already triggered the action multiple times. Can you please take a look into the runners?

**Comment by PSchmiedmayer** [comment]
@philippzagar Thank you for the improvements! I published your 0.4.0 release tag for Spezi Onboarding 👍

Regarding the build errors: I restarted the builds, will check them if they are running now, or I can check the build agents when I am in the office in a few minutes 👍

**Comment by PSchmiedmayer** [comment]
@philippzagar I took a look at the issues. Seems like they were two things happening here:
1. The GitHub Runners are veeeeery slow, and this resulted in some timeouts. Fixed this by switching to our self-hosted runners and adding a section in the README for people to highlight that they might need to modify the CI setup.
2. Seems like the UI tests were failing once I updated all dependencies + set the fixed version for the onboarding flow PR that was tagged in `0.4.0`.
The latest builds are now running, I hope this resolves the issues 👍

**Comment by philippzagar** [comment]
Thanks a lot for putting the time into it! @PSchmiedmayer 🚀

1. Interesting to see that the GitHub runners have that crazy fluctuations in execution speed.. Is it an option to only use our self-hosted runners in order to have more control over this behavior?
2. Thanks for updating the dependencies, especially to the `0.4.0` `SpeziOnboarding` version. Very interesting that the UI tests failed with these updates. The `\n` fix is quite typical and I have encountered it a couple of times already (I think it has something to do with the iOS version 16/17 on the Simulator?). Thanks!

As the tests finally ran through, I would go ahead and merge the PR. I'll draft a new release and you can publish it then 👍

**Comment by PSchmiedmayer** [comment]
Thank you for all the work that went into the PR! 🚀

Yes, the current configuration will enforce only using self-hosted runners for tests from now on.
And yes, I think the issue there is that the text entry already expects a focus while the `enterText` function dismisses focus after text entry. Seems like this is resolved now 👍

**Pull request merged**
